Server Group: add resource and controller

ezeriver94 opened the pull request:
Implementation of Server Groups resource

- create CRD + actuator/controller + e2e tests
- extend Server CRD to support `spec.resource.serverGroupRef` and `status.resource.serverGroups`

some comments:
- `NovaMinimumMicroversion` had to be updated from 2.60 to 2.71 (as it's the first one that supports ServerGroups in Server response). Check [updated comment](https://github.com/ezeriver94/openstack-resource-controller/blob/feat/server-groups/internal/osclients/compute.go#L35) in case the documentation is wrong
- gophercloud's Server createOpts supports setting only one ServerGroup within `SchedulerHintOpts`, but Server object defines `serverGroups` as an array. This implementations reflects exactly what gophercloud (and nova API) exposes, although it could make sense to change `server.status.resource.serverGroups` to `serverGroupID` as only one group is allowed
- ServerGroup list can only filter by name, this means e2e test case `import-error` does not apply to this resource as there can't be more than 1 different group with the same name. This was explained in [README.md](https://github.com/ezeriver94/openstack-resource-controller/blob/feat/server-groups/internal/controllers/servergroup/tests/servergroup-import-error/README.md)
- e2e test case for `server-create-full` was extended with the new ServerGroup field and dependency

mandre commented:
You're on a roll! I'll try to make a first pass in the coming days.

mandre commented on internal/controllers/servergroup/tests/servergroup-create-full/00-servergroup.yaml:
```diff
@@ -0,0 +1,15 @@
+---
```
Can you rename this file to `00-create-resource.yaml` like I did in https://github.com/k-orc/openstack-resource-controller/pull/417/, to bring more consistency across tests?

mandre commented on internal/controllers/servergroup/tests/servergroup-create-minimal/00-servergroup.yaml:
```diff
@@ -0,0 +1,12 @@
+---
```
ditto on the file renaming.

ezeriver94 commented on api/v1alpha1/servergroup_types.go:
```diff
@@ -0,0 +1,93 @@
+	ServerGroupPolicySoftAntiAffinity ServerGroupPolicy = "soft-anti-affinity"
+)
+
+type ServerGroupRules struct {
```
Hmm I'm not sure I understand, as Rules is a struct and not an enum. If I understand correctly there could be room for more properties in the same instance (eventually)

ezeriver94 commented on internal/osclients/compute.go:
```diff
@@ -32,16 +32,18 @@ import (
-NovaMinimumMicroversion is the minimum Nova microversion supported by CAPO
-2.60 corresponds to OpenStack Queens
+NovaMinimumMicroversion is the minimum Nova microversion supported by CAPO and ORC.
+2.71 corresponds to OpenStack Stein
```
Agree, I would leave that part to you 😄

ezeriver94 commented:
Hi @mandre, thanks again for the review. I believe I've addressed all comments, I kept the new commit separated from the main one in case you want to take a closer look to the new commit. Once you think it's ready for merging I will squash

mandre commented:
I'm happy with your changes, squash the commits whenever you want and we'll merge it. You might want to reference the related issue in the commit message or in the PR description (Fixes #402) so that it closes the issue directly and links to the PR from it.
